Dockerfile

```diff
@@ -1,17 +1,34 @@
-FROM node:12-alpine
-
+FROM node:12-alpine as base
 WORKDIR /app
-
 COPY package*.json ./
+ENV PORT=8080
 
-RUN npm install
-
+FROM base AS dev
+RUN npm ci
 COPY . .
+CMD ["npm", "run", "dev"]
 
+FROM dev AS test
+COPY . .
+CMD ["npm", "run", "test"]
+
+# EXPOSE 8080
+FROM base AS builder
+RUN npm ci
+COPY src src
+COPY webpack* .
+COPY .eslintrc.json .
+COPY .babelrc .
+COPY tsconfig.json .
 RUN npm run build
 
-ENV PORT=8080
-
-EXPOSE 8080
-
+FROM base AS prod
+# this still installs all the non-dev dependencies; removing react etc would be better
+RUN npm ci --production
```

Comment:
If you were to run this as an express server, the cleanest way would be to put everything in devDependencies so this only installs express. That's kind of bad coding practices though so it's a pick your poison thing. These images are already way smaller when the devDependencies are pruned so it's not like we have a lack of space right now.